lib/plugins/bull-mq-metrics/MetricsCollector.ts

```diff
@@ -0,0 +1,178 @@
+      queueDiscoverer: new BackgroundJobsBasedQueueDiscoverer(this.redis),
+      ...options,
+    }
+
+    this.metrics = this.registerMetrics(this.registry, this.options)
+  }
+
+  async start() {
+    const queueNames = await this.options.queueDiscoverer.discoverQueues()
+
+    // `void` is used to run the `observeQueue` function without waiting for it to finish
+    void Promise.all(
+      queueNames.map((name) => this.observeQueue(name, this.redis, this.metrics, this.options)),
+    )
+  }
+
+  stop() {
+    this.active = false
+  }
+
+  private registerMetrics(
+    registry: prometheus.Registry,
+    { metricsPrefix, histogramBuckets }: MetricCollectorOptions,
+  ): Metrics {
+    const metrics = getMetrics(metricsPrefix, histogramBuckets)
+    const metricNames = Object.keys(metrics)
+
+    // If metrics are already registered, just return them to avoid triggering a Prometheus error
+    if (metricNames.length > 0 && registry.getSingleMetric(metricNames[0])) {
+      const retrievedMetrics = registry.getMetricsAsArray()
+      const returnValue: Record<string, prometheus.MetricObject> = {}
+
+      for (const metric of retrievedMetrics) {
+        if (metricNames.includes(metric.name)) {
+          returnValue[metric.name as keyof Metrics] = metric
+        }
+      }
+
+      return returnValue as unknown as Metrics
+    }
+
+    for (const metric of Object.values(metrics)) {
+      registry.registerMetric(metric)
+    }
+
+    return metrics
+  }
+
+  private async observeQueue(
+    name: string,
+    redis: Redis,
+    metrics: Metrics,
+    { collectionIntervalInMs }: MetricCollectorOptions,
+  ) {
+    const queue = new Queue(name, { connection: redis })
+    const events = new QueueEvents(name, { connection: redis })
+
+    events.on('completed', (completedJob: { jobId: string }) => {
+      queue
+        .getJob(completedJob.jobId)
+        .then((job) => {
+          if (!job) {
+            return
+          }
+
+          if (job.finishedOn) {
+            metrics.completedDuration
+              .labels({ queue: name })
+              .observe(job.finishedOn - job.timestamp)
+
+            if (job.processedOn) {
+              metrics.processedDuration
+                .labels({ queue: name })
+                .observe(job.finishedOn - job.processedOn)
+            }
+          }
+        })
+        .catch((err) => {
+          this.logger.warn(err)
+        })
+    })
+
+    while (this.active) {
+      const { completed, active, delayed, failed, waiting } = await queue.getJobCounts()
+
+      metrics.activeGauge.labels({ queue: name }).set(active)
+      metrics.completedGauge.labels({ queue: name }).set(completed)
+      metrics.delayedGauge.labels({ queue: name }).set(delayed)
+      metrics.failedGauge.labels({ queue: name }).set(failed)
+      metrics.waitingGauge.labels({ queue: name }).set(waiting)
+
+      await setTimeout(collectionIntervalInMs)
```

Comment:
Do we want to introduce a separate scheduling logic? Doesn't that make things more complicated for our case?

Reply:
Ideally, tests should run with `monitoringEnabled: false`

Reply:
It would allow us to handle all the scheduling logic consistently - either through BullMQ jobs, or through `fastify-schedule`. Adding non-observable, non-exposed, non-transparent `setTimeout` to the mix is not very flexible. And if someone wants it for whatever reason, they could just invoke the collection method inside a `setInterval` on the app startup.

Reply:
if we absolutely want to make this as seamless as possible, I'd suggest requiring fastify-schedule plugin to be registered and use that

Reply:
ah, btw, one extra reason why we may not want to just run this in-memory from the plugin - locking.

If you are running 5 instances of the service, you probably don't need to collect this data 5 times. It would be sufficient if just one service does it; but for that you need to do some kind of concurrency control (via bullmq) or locking (via redis-semaphore)